libp2p/protocols/kademlia/kademlia.nim

```diff
@@ -68,6 +77,66 @@ proc waitRepliesOrTimeouts(
 
   return (receivedReplies, failedPeers)
 
+proc dispatchPutVal(
+    kad: KadDHT, peer: PeerId, entry: ValidatedEntry
+): Future[void] {.async: (raises: [CancelledError, DialFailedError, LPStreamError]).} =
+  let conn = await kad.switch.dial(peer, KadCodec)
+  defer:
+    await conn.close()
+
+  let msg = Message(
+    msgType: MessageType.putValue,
+    record: some(Record(key: some(entry.key.data), value: some(entry.value.data))),
+  )
+  await conn.writeLp(msg.encode().buffer)
+
+  let reply = Message.decode(await conn.readLp(MaxMsgSize)).valueOr:
+    # todo log this more meaningfully
+    error "putValue reply decode fail", error = error
+    return
+
+  if reply != msg:
+    error "unexpected change between msg and reply: ", msg = msg, reply = reply
+
+proc putValue*(
+    kad: KadDHT, key: keys.Key, value: EntryVal, timeout: Option[int]
+): Future[Result[void, string]] {.async: (raises: [CancelledError]), gcsafe.} =
+  let cand = EntryCandidate(key: EntryKey(data: key.getBytes()), value: value)
+  if not kad.entryValidator.isValid(cand):
+    return err("invalid key/value pair")
+  let ts = TimeStamp(ts: $times.now().utc)
+  try:
+    let others: seq[RecordVal] =
+      if cand.key in kad.dataTable.entries:
+        @[kad.dataTable.entries[cand.key]]
+      else:
+        @[]
+
+    let candAsRec = RecordVal(value: cand.value, time: ts)
+    let confirmedRec = kad.entrySelector.select(candAsRec, others)
+
+    let confirmedEnt = EntryCandidate(key: cand.key, value: confirmedRec.value)
+    if not kad.entryValidator.isValid(confirmedEnt):
+      return err("invalid overruled entry")
+    let validEnt = ValidatedEntry.take(confirmedEnt)
+
+    let peers = await kad.findNode(key)
+    var ammo: seq[PeerId]
+    for p in peers.filterIt(it.data != kad.rtable.selfId.getBytes()):
+      ammo.add(p)
+      if ammo.len >= DefaultReplic:
+        break
+    let putClip = ammo.mapIt(kad.dispatchPutVal(it, validEnt))
+    kad.dataTable.insert(validEnt, ts)
+    try:
+      await putClip.allFutures().wait(chronos.seconds(timeout.get(5)))
+    except AsyncTimeoutError:
+      discard
+
+    return results.ok()
+  except:
+    return err("todo: refine exceptions")
```

Comment:
Since this will catch all the exceptions, could we in the meantime add the exception message to the `err` so at during development we notice the type of exceptions happening here?